### PR TITLE
Optimize DoFileSearch on all platforms

### PR DESCRIPTION
**Pull request by JosJuice**
My main reason for doing this was to avoid the slow function File::ScanDirectoryTree which previously was used on all OSes other than Windows, but it seems like this improved the speed a bit on Windows too; the DoFileSearch call that happens when pressing the game list refresh button took about 50-80 ms before this and 25-50 ms after this. (The directories that I tested with mostly contain files with non-matching extensions.) The performance improvement on other platforms should be much larger because of how slow File::ScanDirectoryTree was.

TODO:
- [x] Debug ScanDirectoryTree (it seems to act as if the directory it tries to scan doesn't exist)
- [x] Test on operating systems other than Windows

### CODE REVIEW

**Comment by InusualZ** [comment]
Since you guys updated to VS 2017, why don't use the new `std::filesystem` ?

**Comment by JosJuice** [comment]
We still target non-Windows compilers that don't support `std::filesystem`, so even if we used it, we would still end up needing to have different code for Windows and non-Windows. I also don't like relying on something that VS only has in an `experimental` namespace. And like the PR description says, using `std::filesystem` does seem to be slower than this PR's solution – that's not a good reason for not using `std::filesystem` at all, but it's a reason to consider not using it for game list scanning.

**Comment by leoetlino** [comment]
0xc00000fd apparently indicates a stack overflow or some failed assertion.

**Comment by JosJuice** [comment]
I found the bug that was causing the stack overflows in Windows debug builds and seemingly non-existent directories in Windows release builds: The way that FileUtil.cpp was creating Directory objects led to dangling string references. I've fixed it by moving the ToOSEncoding call into Directory.

That the Windows debug buildbot failed when running tests turned out to be useful for finding the bug :)

**Comment by bb010g** [comment]
Needs a rebase.

File: Source/Core/Common/DirectoryIterator.cpp [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+#include <dirent.h>
+#include "Common/FileUtil.h"
+#endif
+
+#include "Common/StringUtil.h"
+
+namespace Common
+{
+DirectoryEntry::DirectoryEntry()
+{
```
Comment (Warepire) [inline Source/Core/Common/DirectoryIterator.cpp]: This can likely be ` = default;` instead.

File: Source/Core/Common/DirectoryIterator.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  u64 GetSize() const;
+
+private:
+  void Close();
+  void Move(DirectoryEntry&& other);
+
+  void SkipSpecialName();
+
+#ifdef _WIN32
+  HANDLE m_handle = INVALID_HANDLE_VALUE;
```
Comment (Warepire) [inline Source/Core/Common/DirectoryIterator.h]: Could you get rid of enough calls to Close() if you turned Close() into a deleter and the handle into a unique_ptr for it to be worth it in regards to the more complicated calls to the traversal methods?

Comment (JosJuice) [reply Source/Core/Common/DirectoryIterator.h]: We only have two calls to `Close`, so I don't think it's worth it.

File: Source/Core/Common/DirectoryIterator.cpp (hunk 1 above)
Comment (JosJuice) [reply Source/Core/Common/DirectoryIterator.cpp]: Done.

**Comment by JosJuice** [comment]
Rebased. If you want to test this, keep in mind that DolphinQt2's game list scanning code doesn't use DoFileSearch. It is however still used by various functionality as well as DolphinWX's game list scanning, and the Android game list scanning started using it recently.

**Comment by JosJuice** [comment]
Like I mentioned in my previous comment, DolphinQt isn't using this, so the performance improvements aren't really important anymore. (Yes, it would speed up game list scanning on Android, but people generally don't have very big game libraries there...)

While I still think this PR would be nice for cleaning up the code in FileUtil.cpp, this is basically a stopgap solution for until we can get `std::filesystem` on all platforms. And since people don't seem very enthusiastic about this PR (presumably due to added complexity), I'll be closing it now.

**Pull request closed without merging**